src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1162 @@
+                addedAttributes.set(UnsignedInt(mapping.attrib), true);
+                if (mapping.type == MaterialAttributeType::Float) {
+                    Float value = Float(colorMap.value_real) * factor;
+                    arrayAppend(attributes, {mapping.attrib, value});
+                } else if (mapping.type == MaterialAttributeType::Vector3) {
+                    Vector3 value = Vector3(colorMap.value_vec3) * factor;
+                    arrayAppend(attributes, {mapping.attrib, value});
+                } else if (mapping.type == MaterialAttributeType::Vector4) {
+                    Vector4 value = Vector4(colorMap.value_vec4) * Vector4{factor,factor,factor,1.0f};
+                    arrayAppend(attributes, {mapping.attrib, value});
+                } else {
+                    CORRADE_INTERNAL_ASSERT_UNREACHABLE(); /* LCOV_EXCL_LINE */
+                }
+            }
+
+            if (mapping.texture != MaterialAttribute{}) {
+                ufbx_texture *texture = colorMap.texture;
+                if (texture && colorMap.texture_enabled && !addedAttributes[UnsignedInt(mapping.texture)]) {
+                    addedAttributes.set(UnsignedInt(mapping.texture), true);
+                    arrayAppend(attributes, {mapping.texture, UnsignedInt(texture->typed_id)});
+
+                    if (mapping.textureMatrix != MaterialAttribute{} && texture->has_uv_transform) {
+
+                        /* Texture matrix should be connected to texture */
+                        CORRADE_INTERNAL_ASSERT(!addedAttributes[UnsignedInt(mapping.textureMatrix)]);
+                        addedAttributes.set(UnsignedInt(mapping.textureMatrix), true);
+
+                        const ufbx_matrix &mat = texture->uv_to_texture;
+                        Matrix3 value = {
+                            { Float(mat.m00), Float(mat.m10), 0.0f },
+                            { Float(mat.m01), Float(mat.m11), 0.0f },
+                            { Float(mat.m03), Float(mat.m13), 1.0f },
+                        };
+
+                        arrayAppend(attributes, {mapping.textureMatrix, value});
+                    }
+
+                    /* @todo Texture coordinates, either in ufbx or here */
+                }
+            }
+        }
+    }
+
+    if (material->features.double_sided.enabled) {
+        arrayAppend(attributes, {MaterialAttribute::DoubleSided, true});
+
+    }
+
+    /* Can't use growable deleters in a plugin, convert back to the default
+       deleter */
+    arrayShrink(attributes, DefaultInit);
+
+    MaterialTypes types;
+    types |= MaterialType::Phong;
+    if (material->features.metalness.enabled) {
+        types |= MaterialType::PbrMetallicRoughness;
+    }
+
+    Containers::Array<UnsignedInt> layers{ 1 };
+    layers.back() = UnsignedInt(attributes.size());
+
+    return MaterialData{types, std::move(attributes), std::move(layers)};
```

Comment:
FBX does not as far as I know support layered materials (outside of some extremely cursed stuff in 3ds Max, but that's far from *standard* FBX). It does support layered *textures* so those could be exposed at least as layers. I'll have to take a quick peek at what Assimp considers layers though.

Reply:
Looked into it: Seems like Assimp just interprets the layered textures as layered materials, do we want to do a similar thing with ufbx?

Reply:
Oh also, FBX represents texture layer binding with named UV sets but Magnum expects them to be indices. I could try to match them up depending on what meshes a texture is bound to, it should be safe in most cases but there's a risk of one texture being bound to multiple meshes with different order of UV set names, but I guess I could just omit the material attribute in such cases.

Reply:
The layers were made specifically to serve that use case -- by design it's not possible to have, say, two `DiffuseTexture` entries in a single material layer because it would be unclear how those should be interpreted. Each has to be in its own layer, together with the layer containing some kind of info about how the texture is meant to be applied:

- either implicitly (e.g. `Roughness` in the `ClearCoat` layer has a well-defined way of what is it used for, in contrast to `Roughness` in the base metallic/roughness material) 
- or explicitly via some (custom) attribute (which I assume the FBX material defines in some way -- it was created using [this tutorial](https://help.autodesk.com/view/FBX/2017/ENU/?guid=__files_GUID_452768B5_7A4F_45BD_AC37_DC1A034DCF3B_htm) and I see `layeredTexture.SetTextureBlendMode( i, fbx.FbxLayeredTexture.eAdditive )` there, not sure what exactly that maps to in the file itself).

Reply:
Should I extract the PBR layers like coat/sheen to their own layers even though they're not layered textures in FBX? (Hmm just found `MaterialLayer::ClearCoat` as I was writing this so I suppose that's a yes :D)

And for FBX layered textures they could have different blend modes for different textures, so I guess we'd have something like `diffuseTextureBlend` or `specularTextureBlend`.

Reply:
> `MaterialLayer::ClearCoat` [...]

Yes :) I might reconsider particular cases once I get to investigating & integrating more PBR properties (e.g. glTF's [KHR_materials_specular](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_specular/README.md) probably will make more sense being in the base material than in a layer), but it won't hurt if you make them all layers.

> different blend modes for different textures, so I guess we'd have something like `diffuseTextureBlend` or `specularTextureBlend`

Yup!